osu.Game/Database/RealmFileStore.cs

```diff
@@ -50,16 +53,23 @@ public RealmFile Add(Stream data, Realm realm, bool addToRealm = true)
             var file = existing ?? new RealmFile { Hash = hash };
 
             if (!checkFileExistsAndMatchesHash(file))
-                copyToStore(file, data);
+                copyToStore(file, data, preferHardLinks);
 
             if (addToRealm && !file.IsManaged)
                 realm.Add(file);
 
             return file;
         }
 
-        private void copyToStore(RealmFile file, Stream data)
+        private void copyToStore(RealmFile file, Stream data, bool preferHardLinks)
         {
+            if (RuntimeInfo.OS == RuntimeInfo.Platform.Windows && data is FileStream fs && preferHardLinks)
+            {
+                // attempt to do a fast hard link rather than copy.
+                if (HardLinkHelper.CreateHardLink(Storage.GetFullPath(file.GetStoragePath()), fs.Name, IntPtr.Zero))
```

Comment:
Ok well I was testing some more and then noticed that hard links actually *weren't* getting created sometimes. Reproduction steps:

1. Go to `%appdata%` and nuke entire `osu-development-1` folder (if exists)
2. Launch game with debug in `osu! (Second Client)` config, breakpoint here
3. Go to first set up overlay, attempt to import from stable
4. This call returns `false` almost every time, and `Marshal.GetLastWin32Error` returns 3, which is _The system cannot find the path specified._

The reason for this failure that when a file store is started on a brand new install, its hashed directory structure may not be initialised yet. So the path to create the target hard link at partially does not exist because the hashed directories don't.

Fix is easy, thankfully:

```suggestion
                if (HardLinkHelper.CreateHardLink(Storage.GetFullPath(file.GetStoragePath(), true), fs.Name, IntPtr.Zero))
```

This ensures that `Storage.GetFullPath()` creates the directories, eliminating this failure vector.

Reply:
Great catch. I was testing with a fully populated cache directory structure so didn't even notice this..